File: catalog/segument_tree/segment_tree.js
"use strict"

class SegmentTree {
  constructor(xs) {
    const n = xs.length
    this.n = n

    this.nodes = new Array(2*n - 1)
    const m = this.nodes.length

    // 葉に値をセット
    for (var i = 1; i <= n; i++) {
      this.nodes[m-i] = xs[n-i]
    }

    // 後ろから順に値をセット
    for (var k = m - n - 1; k >= 0; k--) {
      // 左の子
      const left = this.nodes[2 * k + 1]
      // 右の子
      const right = this.nodes[2 * k + 2]
      // 自分の値
      this.nodes[k] = Math.min(left, right)
    }
  }
  update(i, x) {
    // 葉の段にアクセス
    const start = (this.nodes.length - n)
    const k = start + i
    this.nodes[k] = x
    for (var j = Math.floor((k-1)/2); j >= 0; j = Math.floor((j-1)/2)) {
      // 左の子
      const left = this.nodes[2 * j + 1]
      // 右の子
      const right = this.nodes[2 * j + 2]
      // 自分の値
      const v = Math.min(left, right)
      this.nodes[j] = v
    }
  }
  query(a, b) {
    const start = this.nodes.length - n
    return this._query(start+a, start+b, 0, start, this.nodes.length)
  }
  _query(a, b, k, l, r) {
    // [a, b) と [l, r] が関係ない
    if (b <= l || a >= r) return Number.MAX_SAFE_INTEGER
    // [a, b) 内に収まってる
    else if (a <= l && r <= b) return this.nodes[k]
    // [a, b) と [l, r] が一部交わってる
    else {
      return Math.min(
        this._query(a, b, 2*k + 1, l, Math.floor((r+l)/2)),
        this._query(a, b, 2*k + 2, Math.floor((r+l)/2), r)
      )
    }
  }
}

/**
 * % node main.js
 * > 4
 * > 5
 * > 3
 * > 7
 * > 9
 *   xs:5,3,7,9
 *   SegmentTree { n: 4, nodes: [
 *       3, 3, 7, 5,
 *       3, 7, 9
 *     ] }
 *   update i=1 x=9
 *   SegmentTree { n: 4, nodes: [
 *       5, 5, 7, 5,
 *       9, 7, 9
 *     ] }
 *   query [0, 3]
 *   5
 */
let count = 0
let n = 0
const xs = []
const reader = require("readline").createInterface({
  input: process.stdin,
  output: process.stdout,
})
reader.on("line", (line) => {
  if (count === 0) {
    n = Number(line.trim())
  } else {
    const x = Number(line.trim())
    xs.push(x)
    if (count === n) {
      console.log(`xs:${xs}`)
      const tree = new SegmentTree(xs)
      console.log(tree)
      console.log(`update i=1 x=9`)
      tree.update(1, 9)
      console.log(tree)
      const min = tree.query(0, 3)
      console.log(`query [0, 3]`)
      console.log(min)
      process.exit(0)
    }
  }
  count++
})

File: catalog/segument_tree/segment_tree.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

type segmentTree struct {
	nodes []int
	n     int
}

func newSegmentTree(size int, xs []int) *segmentTree {
	tree := &segmentTree{}

	// 葉の数
	tree.n = size

	// 全ノード数
	tree.nodes = make([]int, 2*size-1)

	// 葉をセット
	start := len(tree.nodes) - size
	for i := 0; i < size; i++ {
		tree.nodes[start+i] = xs[i]
	}

	// 登っていきながら値をセット
	for k := start - 1; k >= 0; k-- {
		left := tree.nodes[2*k+1]
		right := tree.nodes[2*k+2]
		if left > right {
			tree.nodes[k] = right
		} else {
			tree.nodes[k] = left
		}
	}

	return tree
}

func (t *segmentTree) update(i int, x int) {
	start := len(t.nodes) - t.n + i
	t.nodes[start] = x

	// 親
	k := start
	for k > 0 {
		k = (k - 1) / 2
		left := t.nodes[2*k+1]
		right := t.nodes[2*k+2]
		if left > right {
			t.nodes[k] = right
		} else {
			t.nodes[k] = left
		}
	}
}

func (t *segmentTree) query(a int, b int) int {
	start := len(t.nodes) - t.n
	return t._query(start+a, start+b, 0, start, start+t.n-1)
}

func (t *segmentTree) _query(a int, b int, k int, left int, right int) int {
	if b <= left || right <= a {
		return math.MaxInt32
	} else if a <= left && right <= b {
		return t.nodes[k]
	} else {
		l := t._query(a, b, 2*k+1, left, (left+right)/2)
		r := t._query(a, b, 2*k+2, (left+right)/2, right)
		if l < r {
			return l
		} else {
			return r
		}
	}
}

func (t *segmentTree) String() string {
	return strings.Trim(fmt.Sprint(t.nodes), "[]")
}

func nextInt(sc *bufio.Scanner) int {
	sc.Scan()
	s := sc.Text()
	x, err := strconv.Atoi(s)
	if err != nil {
		panic(err)
	}
	return x
}

/**
 * $ go run segment_tree.go
 * > 8
 * > 5 3 7 9 1 4 6 2
 * 1 3 1 3 7 1 2 5 3 7 9 1 4 6 2
 * 2 3 2 3 7 4 2 5 3 7 9 10 4 6 2
 * 2
 */
func main() {
	sc := bufio.NewScanner(os.Stdin)
	sc.Split(bufio.ScanWords)

	n := nextInt(sc)
	xs := make([]int, n)
	for i := 0; i < n; i++ {
		xs[i] = nextInt(sc)
	}

	tree := newSegmentTree(n, xs)
	fmt.Println(tree)
	tree.update(4, 10)
	fmt.Println(tree)
	fmt.Println(tree.query(0, 7))
}
